### ISSUE TITLE
ZSWAP is not available in the nix kernels

### ISSUE THREAD

**Issue by CMCDragonkai** [NixOS/nixpkgs#11596]
I added `zswap.enabled=1` for the kernel command line:

```
cat /proc/cmdline
initrd=... zswap.enabled=1
```

But `echo 1 > /sys/module/zswap/parameters/enabled` doesn't work. There's no zwap module.

Then I looked for the kernel config: `cat /proc/config.gz | gunzip | grep ZSWAP` nothing.

ZSWAP should be available by 3.18.24 (the kernel I'm running), so any reason why zswap wasn't compiled into the kernel?

**Comment by zimbatm** [comment]
Link to relevant do: https://wiki.archlinux.org/index.php/Zswap
Grepping trough nixpkgs doesn't reveal anything either

**Comment by nckx** [comment]
Is simply compiling a custom kernel with `ZSWAP_... y` in `kernelExtraConfig` not an option?

Anecdotal experience: zswap can be quite unstable even on release kernels (4.3, here). I'd be weary of building it in by default.

**Comment by vcunat** [comment]
Building it as a module shouldn't cause problems to those not loading it explicitly, should it?

**Comment by nckx** [comment]
Ah, no comment. I've never done that since having other problems with it modularised (of the ‘it just doesn't work’ variety, not the flaming data death kind).

But that was a very long time ago, so it's probably fine. The aforementioned panics were not (4.3, I think?).

**Comment by nckx** [comment]
I must have been thinking of zsomethingelse, as was all the rage around 3.0 (frontswap? cleancache? I really can't remember). `CONFIG_ZSWAP` would be boolean.

And `kernelExtraConfig` the way to go.

**Comment by abbradar** [comment]
Closed by https://github.com/NixOS/nixpkgs/commit/7b9c493d6020d29b2229fffad35acf99c3b47a61

**Comment by jerith666** [comment]
I just merged in `nixos-unstable` containing 7b9c493, and now I have:

```
$ journalctl -b | grep zswap
Jul 24 14:36:46 alpha kernel: zswap: default zpool zbud not available
Jul 24 14:36:46 alpha kernel: zswap: pool creation failed
```
